### DISCUSSION TITLE
Extensive integration tests

### DISCUSSION THREAD

**Discussion by iboB** [cpm-cmake/CPM.cmake#224] [Ideas]
I think we need more robust and extensive integration testing so we can test new features and especially regressions.

So my idea is to basically have `test/integration` which includes several projects with multiple dependencies with multiple configurations (with and without package-lock, external cache), some projects with errors. These dependencies could be existing GitHub repos, but I think that having multiple (perhaps 10-15) repos in `cpm-cmake` named `test-package-<something>` will give us more control over what exactly is tested. These test repos can provide versions, archives, themselves have interdependencies and more. 

The thing is I don't think that driving the integration tests with CMake is a good idea. CMake is just not a very nice language for such things, and while possible, I think using Python, or Ruby, or another, more general purpose language, will make things easier and much more pleasant, and crucially help us minimize bugs in the tests themselves.

The integration test projects may also serve as examples of advanced usage of CPM

**Comment by Andres6936** [comment]
I recently managed to run the [test suite](https://github.com/Andres6936/Cup.Graphics/runs/1939507256?check_suite_focus=true) for [my project](https://github.com/Andres6936/Cup.Graphics) within Github Actions using Doctest, I [used CPM](https://github.com/Andres6936/Cup.Graphics/blob/b28dff4f74d655903e4b8b9a4bdcc94759d18141/CMakeLists.txt#L14) for downloading and configuring the dependency, it also works using Catch2 (although I don't have the evidence).

**Comment by TheLartians** [comment]
I currently use a rather [large project](https://maphi.app) with ~40 total dependencies (many of them are nested) myself, which I often use as a "works on my machine" integration test before releases. Unfortunately, due to the closed-source nature it can't be used as a public integration test in CI, but some test repos with some interdependent CPM.cmake dependencies and using the full feature scope would be great to have!

I agree that CMake isn't the best test runner for integration tests. It's still great for unit tests, as it allows to call functions and inspect local variables directly, but integration tests are quite clumsy there. Note that the examples (which already serve as small integration tests) are also using a python script for building.

**Comment by iboB** [comment]
@TheLartians how do you feel about the ruby programming language? 😄 

I'm thinking of starting the initial work on the integration tests and to me using ruby as the driver feels most natural. I have extensive experience with it and would definitely work fastest writing in ruby.

Basically the integration test driver's job would be to:

* Generate CMakeLists.txt for root projects
* Configure with various command-line options and env vars (say CPM_SOURCE_CACHE)
* Parse the CMake output for specific expected failures (not just test that if fails, but test that it fails in with a certain error)
* Read the generated CMakeCache.txt for certain values
* Read package lock when applicable and check for expected output
* Run builds
* Check build output for the existence of expected targets (for example that the examples or tests of a depenency are **not** there if we expect them not to be there)
* Other things as they come up

PS Perhaps this discussion should be turned into an issue, what do you think?

**Reply by TheLartians** [reply]
TBH I'm more fluent with Python and JavaScript than Ruby, but if you feel that Ruby is the most natural fit for such things, I'll trust you and try to catch up! 😄

Sure I think tracking progress in an issue makes sense at this point, can we transform the discussion back to issues though?
Otherwise feel free to open one.

**Reply by iboB** [reply]
Oops... you're right. It seems this is not possible. Ok. So I'll add some stuff in a branch in the next couple of days and figure out what do to about issues later
